core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala

```diff
@@ -0,0 +1,89 @@
+package org.broadinstitute.dsde.rawls.snapshot
+
+import bio.terra.datarepo.model.{SnapshotModel, CloudPlatform => SnapshotCloudPlatform}
+import org.broadinstitute.dsde.rawls.RawlsException
+import org.broadinstitute.dsde.rawls.model.WorkspaceCloudPlatform.WorkspaceCloudPlatform
+import org.broadinstitute.dsde.rawls.model.{Workspace, WorkspaceCloudPlatform}
+
+import scala.jdk.CollectionConverters.CollectionHasAsScala
+
+object SnapshotReferenceCreationValidator {
+  def constructor(workspaceContext: Workspace, snapshot: SnapshotModel): SnapshotReferenceCreationValidator =
+    new SnapshotReferenceCreationValidator(workspaceContext, snapshot)
+}
+
+// thrown to disallow creation of a snapshot reference on an unsupported platform
+class UnsupportedPlatformException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across cloud boundaries
+class PlatformBoundaryException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across a protected data boundary
+class ProtectedDataException(message: String) extends RawlsException(message)
+
+class SnapshotReferenceCreationValidator(val workspaceContext: Workspace, val snapshot: SnapshotModel) {
+
+  // TODO: get this information from a more authoritative source rather than relying on the hardcoded bucket prefix
+  private def isWorkspaceProtected: Boolean = workspaceContext.bucketName.startsWith("fc-secure")
+
+  private def isSnapshotProtected: Boolean = snapshot.getSource.asScala.exists(_.getDataset.isSecureMonitoringEnabled)
+
+  private def snapshotPlatform: SnapshotCloudPlatform = {
+    val snapshotSourceDatasets = snapshot.getSource.asScala
+    val datasetCloudPlatforms = snapshotSourceDatasets.map(_.getDataset.getCloudPlatform).filter(_ != null).toSet
```

Comment:
what you have here is fine; to be Scala-idiomatic, you'd work with `Option` instead of `null`, like:
```
val datasetCloudPlatforms = snapshotSourceDatasets.flatMap(x => Option(x.getDataset.getCloudPlatform)).toSet
```
(boy, I hope that's functionally equivalent; I didn't actually try it)

Reply:
Love it!  Looks like it did the trick.